### PR TITLE
[ENHANCEMENT] Improve the migration

### PR DESCRIPTION
**Pull request by Nexucis**
The main idea is to offload as much as possible of the migration logic to Golang, and delegate to Cuelang only when necessary (= at plugin model level), in order to solve #2039.

That's a big change so let's keep an eye on that, but tests are passing & memory usage + computing time is much much less than it was.

### CODE REVIEW

**Comment by AntoineThebaud** [comment]
https://github.com/perses/perses/blob/main/scripts/check_license.sh#L6 "migrate.cue" can be removed there

File: cue/schemas/panels/bar/migrate.cue [hunk 1]
```diff
@@ -1,16 +1,36 @@
...
-		#decimal: *#panel.fieldConfig.defaults.decimal | null
-		if #decimal != null {
-			format: decimalPlaces: #decimal
-		}
+import (
+	commonMigrate "github.com/perses/perses/cue/schemas/common/migrate"
+)
+
+#grafanaType: "bargauge"
+#panel:       _
```
Comment (AntoineThebaud) [inline cue/schemas/panels/bar/migrate.cue]: technically would be more accurate to put 
```suggestion
#panel:       {...}
```

File: internal/api/impl/migrate/migrate.go [hunk 2]
```diff
@@ -48,8 +49,12 @@ func (e *endpoint) Migrate(ctx echo.Context) error {
 	if err := ctx.Bind(body); err != nil {
 		return apiinterface.HandleBadRequestError(err.Error())
 	}
-	grafanaDashboard := migrate.ReplaceInputValue(body.Input, string(body.GrafanaDashboard))
-	persesDashboard, err := e.migrationService.Migrate([]byte(grafanaDashboard))
+	rowGrafanaDashboard := []byte(migrate.ReplaceInputValue(body.Input, string(body.GrafanaDashboard)))
```
Comment (AntoineThebaud) [inline internal/api/impl/migrate/migrate.go:52]:
```suggestion
	rawGrafanaDashboard := []byte(migrate.ReplaceInputValue(body.Input, string(body.GrafanaDashboard)))
```

😉

**Comment by Nexucis** [comment]
PR is now ready. After running this PR on the different dashboards community provides that was showing the memory issue, I am not able to reproduce the issue.

Let's see if on a long running memory is kept at a low level.

File: internal/api/migrate/grafana.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+	type plain SimplifiedDashboard
+	if err := json.Unmarshal(data, (*plain)(tmp)); err != nil {
+		return err
+	}
+	tmp.rearrangeGrafanaPanelsWithinExpandedRows()
+	*d = *tmp
+	return nil
+}
+
+func (d *SimplifiedDashboard) rearrangeGrafanaPanelsWithinExpandedRows() {
```
Comment (AntoineThebaud) [inline internal/api/migrate/grafana.go:188]: let's keep the intro comment too?

```suggestion
/**
 * This function addresses an issue we have with Grafana datamodel when it comes to migrating dashboards to Perses: When
 * a row is expanded in Grafana, its children panels are moved up in the main panels list, thus become siblings of the row.
 * When it comes to Perses migration we need to recompose the parent->children relationships. However in its current state
 * the CUE language doesn't permit us to achieve this recomposition, hence this processing in the backend code.
 *
 * So what this function does is basically the following: whenever such pattern is encountered in the panels list:
 * ...
 * row1,
 * panelA,
 * panelB,
 * panelC,
 * row2,
 * ...
 * the objects get rearranged like:
 * ...
 * row1: {
 *   panelA,
 *   panelB,
 *   panelC,
 * },
 * row2,
 * ...
 */
func (d *SimplifiedDashboard) rearrangeGrafanaPanelsWithinExpandedRows() {
```

File: internal/api/migrate/loader.go [hunk 4]
```diff
@@ -14,112 +14,89 @@
...
 	for _, file := range files {
 		if !file.IsDir() {
 			logrus.Tracef("file %s is ignored since we are looking for directories", file.Name())
 			continue
 		}
-		migFilePath := filepath.Join(c.schemasPath, file.Name(), "migrate.cue")
-		contentStr, readErr := os.ReadFile(migFilePath)
-		if readErr != nil {
-			logrus.WithError(readErr).Debugf("No migration file found at %s, plugin %s will be skipped", migFilePath, file.Name())
+		schemaPath := filepath.Join(panelSchemaPath, file.Name(), "migrate")
```
Comment (AntoineThebaud) [inline internal/api/migrate/loader.go:69]: to be adjusted with the new `migrate` folder added

File: internal/api/migrate/migrate.go [hunk 5]
```diff
@@ -81,250 +46,159 @@ func ReplaceInputValue(input map[string]string, grafanaDashboard string) string
...
+	if err := grafanaValue.Validate(cue.Final()); err != nil {
+		logrus.WithError(err).Trace("Unable to wrap the received json into a CUE definition")
+		return nil, true, apiinterface.HandleBadRequestError(err.Error())
+	}
 
-	newGrafanaDashboardRaw, err := json.Marshal(grafanaDashboard)
-	if err != nil {
-		logrus.WithError(err).Error("Unable to marshal the modified dashboard")
-		return grafanaDashboardRaw
+	// Finally, execute the cuelang script with the static mapping and the Grafana Panel as a scope.
```
Comment (AntoineThebaud) [inline internal/api/migrate/migrate.go:182]: wrong/deprecated comment

File: internal/api/migrate/testdata/simple_perses_dashboard.json [hunk 6]
```diff
@@ -452,20 +455,23 @@
...
                 },
                 {
                   "header": "API server",
                   "name": "apiserver"
                 },
                 {
-                  "header": "ValueCustomHeader",
-                  "name": "value"
+                  "hide": true,
+                  "name": "timestamp"
```
Comment (AntoineThebaud) [inline internal/api/migrate/testdata/simple_perses_dashboard.json:470]: weird that the migration is worse here (merged settings getting split), are you sure your maim branch was up to date here?

File: internal/cli/cmd/migrate/migrate.go [hunk 7]
```diff
@@ -131,8 +131,12 @@ func (o *option) onlineExecution(grafanaDashboard json.RawMessage) (*modelV1.Das
 }
 
 func (o *option) offlineExecution(grafanaDashboard json.RawMessage) (*modelV1.Dashboard, error) {
-	finalGrafanaDashboard := migrate.ReplaceInputValue(o.input, string(grafanaDashboard))
-	return o.mig.Migrate([]byte(finalGrafanaDashboard))
+	rowGrafanaDashboard := []byte(migrate.ReplaceInputValue(o.input, string(grafanaDashboard)))
```
Comment (AntoineThebaud) [inline internal/cli/cmd/migrate/migrate.go:134]:
```suggestion
	rawGrafanaDashboard := []byte(migrate.ReplaceInputValue(o.input, string(grafanaDashboard)))
```

File: pkg/model/api/v1/common/jsonref.go [hunk 8]
```diff
@@ -29,7 +29,7 @@ var (
 
 type JSONRef struct {
 	// Ref is the JSON reference.
-	// That's the only thing used during the marshalling / unmarshalling process.
+	// That's the only thing used during the marshaling / unmarshalling process.
```
Comment (AntoineThebaud) [inline pkg/model/api/v1/common/jsonref.go:32]: learned sth today:

> Marshaling is predominantly used in 🇺🇸 American (US) English ( en-US ) while marshalling is predominantly used in 🇬🇧 British English (used in UK/AU/NZ) ( en-GB )

Then I guess you should either change nothing here OR change unmarsha**ll**ing too ;D

**Review by AntoineThebaud — CHANGES_REQUESTED** [review]
Nice! since we are merging this on main directly I guess the PR should also update the plugins docs accordingly following the structure changes ;)

File: internal/api/migrate/testdata/simple_perses_dashboard.json (hunk 6 above)
Comment (Nexucis) [reply internal/api/migrate/testdata/simple_perses_dashboard.json:470]: and it doesn't look like I am outdated regarding the table plugin

File: internal/api/migrate/loader.go (hunk 4 above)
Comment (Nexucis) [reply internal/api/migrate/loader.go:69]: that's already the case ;)

File: internal/api/migrate/loader.go [hunk 9]
```diff
@@ -14,112 +14,89 @@
...
-	var conditions string
-	conditions, successfulLoadsCount, err := c.buildListOfConditions()
-	if err == nil {
-		c.listOfConditions.Store(&conditions)
+	for _, file := range files {
+		if !file.IsDir() {
+			logrus.Tracef("file %s is ignored since we are looking for directories", file.Name())
+			continue
+		}
+		schemaPath := filepath.Join(path, file.Name(), "migrate")
```
Comment (AntoineThebaud) [inline internal/api/migrate/loader.go:43]:
```suggestion
		schemaPath := filepath.Join(path, file.Name(), migrationFolder)
```

File: internal/api/migrate/loader.go (hunk 4 above)
Comment (AntoineThebaud) [inline internal/api/migrate/loader.go:69]:
```suggestion
		schemaPath := filepath.Join(panelSchemaPath, file.Name(), migrationFolder)
```

**Review by AntoineThebaud — APPROVED** [review]
Awesome, 'actually jealous that I didn't have this idea before you. 

🚀

**Pull request merged**